### DISCUSSION TITLE
Issue with autocomplete in the web editor

### DISCUSSION THREAD

**Discussion by colomboe** [eclipse-langium/langium#1364] [Q&A] [answered]
Hi all, first of all I'd like to say thank you for all the work that has been done with Langium, that I find amazing!

I'm writing here because I'm facing a strange issue, and before opening a github issue here I'd like to understand if I'm doing something wrong.

Basically my issue is that, once I implement my custom scope computation, everything works fine inside vscode using the generated extension; instead, in the web editor (build directly form the original project template) I have a weird behaviour where autocomplete works only for odd characters. Here is an example:

https://www.msec.it/odd-example/recording.gif

This is the source code of a little example I've just created to reproduce the issue. The only change I did compared to the template project created with `yo` are the removal of the default validator and the implementation of the grammar and the scope computation.

Here is the live example editor: https://www.msec.it/odd-example/monacoClassic.html
Here is the example code repo: https://github.com/colomboe/odd-example

Any thought about this? It seems quite easy to run into this issue, did anyone already spotted it before?

**Comment by msujew** [comment]
@colomboe Thanks for the kind words :) I can confirm the issue, that's pretty weird.

@kaisalmen Possible that this is a [`monaco-vscode-api`](https://github.com/CodinGame/monaco-vscode-api)/[`monaco-languageclient`](https://github.com/TypeFox/monaco-languageclient/) issue? Since it works fine in vscode, I don't think it's a Langium issue.

**Reply by kaisalmen** [reply]
@colomboe I see it behave normally and then suddenly stop working properly. Try to suppress word-based suggestions and test if it prevents the error or not. Add `wordBasedSuggestions: false` here: https://github.com/colomboe/odd-example/blob/main/static/setupClassic.js#43 
I haven't noticed this behaviour before. Could be something so far undiscovered.

**Reply by colomboe** [reply]
Thank you both @msujew and @kaisalmen! Actually same issue also with this additional setting...

**Reply by kaisalmen** [reply]
@colomboe Thank you. I will see if I can re-produce the problem without bundled code.

**Reply by colomboe** [reply]
Thank you very much! You can start from my example repo on github if it can be useful for you.

**Reply by colomboe** [reply]
@kaisalmen sorry bothering... any news about this issue? Can I help in any way?

**Reply by kaisalmen** [reply]
@colomboe I didn't get to it last week, but it is on my list for this week.

**Reply by kaisalmen** [reply]
Hi @colomboe my apologies for the delay, but I can re-reproduce your issue with your repo even in the bundled case (vite dev mode with some local code changes). I will see now if I can understand where the actual problem comes from and how to fix it.

**Reply by colomboe** [reply]
Hi @kaisalmen! thank you very much for keeping me posted and for your work! Please let me know if I can help you in any possible way!

**Reply by kaisalmen** [reply, marked as answer]
@colomboe this will be fixed with Langium v3. #1334 has fixed the underlying problem. I was apply to reproduce it with larger grammars where the response time by the language server was longer. Once I updated to a next build from Langium v3 containing the fix, the problem is gone.

**Reply by kaisalmen** [reply]
This was problem only surfaced with the web-worker based language server making use of `monaco-editor-wrapper` and I wasn't able to observe it with VSCode itself

**Reply by colomboe** [reply]
@kaisalmen thank you very much! Hope v3 will be available soon 😄 is there any already available documentation or post describing what will change and how to manage the upgrade?

**Reply by msujew** [reply]
A changelog (including a guide on how to navigate the breaking changes) is available over at https://github.com/eclipse-langium/langium/pull/1382. You can already use the `3.0.0-next.cc1a49c` version that we currently test in some of our projects.

**Reply by colomboe** [reply]
@msujew @kaisalmen I've managed to create a simple porting of my project to Langium v3. Everything seems working fine and the issue with the autocomplete is gone! Thank you very much for all your support!

I'm looking forward for the final v3 release 😄
